fe/fe-core/src/main/java/com/starrocks/clone/TabletScheduler.java

```diff
@@ -233,10 +233,17 @@ public Map<Long, PathSlot> getBackendsWorkingSlots() {
 
     /**
      * add a ready-to-be-scheduled tablet to pendingTablets, if it has not being added before.
-     * if force is true, do not check if tablet is already added before.
      */
     public synchronized AddResult addTablet(TabletSchedCtx tablet, boolean force) {
-        if (!force && containsTablet(tablet.getTabletId())) {
+        // Under no circumstance should we repeatedly add a tablet to pending queue
+        // to schedule, because this will break the scheduling logic. Besides, with current design,
+        // we have to maintain the constraint that `allTabletIds = runningTablets + pendingTablets`.
+        // `allTabletIds` and `runningTablets` are defined as unique container, if we schedule a
+        // tablet with different `TabletSchedCtx` at the same time, the reference in `runningTablets`
+        // can be messed up.
+        // `force` here should only mean that we can exceed the size limit of
+        // `pendingTablets` and `runningTablets` if there is too many tablets to schedule.
+        if (pendingTablets.stream().filter(t -> t.getTabletId() == tablet.getTabletId()).count() != 0) {
```

Comment:
a O(n) check will it be a bottleneck?

Reply:
size of `pendingTablets`  is limited(maximum is 2000 by default) and we already have 4 levels of nested loop in `TabletChecker`..., so i think it's ok but need further test.

Reply:
i think we can avoid O(n) search in `pendingTablets` and still make use of `allTabletIds` map, although in a specific situation we should explicitly remove the tablet id from `allTabletIds` first before `addTablet()`, see the function `addBackToPendingTablets()` for details.